source/Octopus.Tentacle/Communications/AutofacServiceFactory.cs

```diff
@@ -17,48 +17,48 @@ public class AutofacServiceFactory : IServiceFactory, IServiceRegistration, IDis
         // Must never be modified as it is required for backwards compatability in BackwardsCompatibleCapabilitiesV2Decorator
         const string TentacleServiceShuttingDownMessage = "The Tentacle service is shutting down and cannot process this request.";
         readonly ILifetimeScope scope;
-        readonly Dictionary<string, Type> serviceTypes = new();
+        readonly Dictionary<string, KnownService> knownServices = new();
 
         public AutofacServiceFactory(ILifetimeScope scope, IEnumerable<IAutofacServiceSource> sources)
         {
             this.scope = scope.BeginLifetimeScope(b =>
             {
-                foreach (var service in sources.SelectMany(x => x.ServiceTypes.EmptyIfNull()))
+                foreach (var knownService in sources.SelectMany(x => x.KnownServices.EmptyIfNull()))
                 {
-                    BuildService(b, service);
+                    BuildService(b, knownService);
                 }
             });
         }
 
-        void BuildService(ContainerBuilder builder, Type serviceType)
+        void BuildService(ContainerBuilder builder, KnownService knownService)
         {
-            var registrationBuilder = builder
-                .RegisterType(serviceType)
+            //register the service implementation AsSelf(), so we can resolve it in the future
+            builder
+                .RegisterType(knownService.ServiceImplementationType)
                 .AsSelf()
                 .SingleInstance();
 
-            var interfaces = serviceType.GetInterfaces();
-            if (serviceType.IsInterface || interfaces.IsNullOrEmpty())
-            {
-                throw new InvalidServiceTypeException(serviceType);
-            }
-            
-            foreach (var face in interfaces)
+            //track the contract type to their known service implementations
+            //the contract type is the one that is sent across the wire (typically an IService sync contract)
+            knownServices[knownService.ServiceContractType.Name] = knownService;
+
+            //the implementation type doesn't need to implement the service contract, but it's _implied_ that it should have _some_ interfaces
+            if (knownService.ServiceImplementationType.IsInterface || knownService.ServiceImplementationType.GetInterfaces().IsNullOrEmpty())
```

Comment:
I feel like this validation would be better placed inside the `KnownService` constructor itself. That way, you can never have an invalid `KnownService`

Reply:
@sburmanoctopus  I've made this change in 1a95c8d9

Reply:
Nice one!

While looking at that, I wonder if we should also validate that `serviceContractType` is an interface?

I don't want to keep moving the goalposts, so feel free to ignore this idea 🙂

Reply:
That's a good idea 👍

Reply:
@sburmanoctopus added expanded validation in 21429d6f and changed the exception type and cleaned up the message

Reply:
Nice, LGTM 🙂